### DISCUSSION TITLE
Feature requests

### DISCUSSION THREAD

**Discussion by khorolets** [near/near-indexer-for-explorer#238] [Ideas]
During the recent issues with 1.24.0-rc.1 release of `nearcore` chain team had to investigate what is going on. Afterward, they said:
> Epochs (and epoch boundaries) are the source of a lot of ‘ifs’ in our code. At the same time, I couldn’t find an easy way to see when given epochs started/ended. **This data was not present in explorer - and was not in indexer database.**

Here's the clarification about what kind of data may be useful. I think we should consider storing it:

- for each block: which epoch it belongs to (and also maybe which validator produced it)
- have a 'list' of epochs (separate table?) with info on the first/last block that belonged to them and which validators participated.
- it would be great to also have some info for block heights that were NOT produced -- for example info on who was supposed to produce it.  (same for chunks)

@frol @telezhnaya what do you think?

**Comment by telezhnaya** [comment]
I think it's a great idea! Though, I'd prefer to clarify whether the users need №3 (we need to think/ask somewhere about the usecases).
Thank you for collecting this!

**Comment by frol** [comment]
> it would be great to also have some info for block heights that were NOT produced -- for example info on who was supposed to produce it. (same for chunks)

We may save epoch summaries at the last block of the epoch (see https://docs.near.org/docs/api/rpc/network#validation-status). This will allow us to track something like https://github.com/frol/near-validators-scoreboard/blob/scoreboard-mainnet/validators_scoreboard.csv

I am not sure it is worth to store the exact block height of the non-produced blocks (I imagine someone would only care about the amount of missing blocks rather than a specific height)

**Comment by frol** [comment]
> for each block: which epoch it belongs to

Given that we always have 12*60*60 blocks with the same epoch, I am concerned about storing this much of duplicate information in the database. Having a separate table with the epoch summary (epoch id, start block height, end block height, rewards, number of missed blocks, number of missed chunks, ???) seems to be a better option

> (and also maybe which validator produced it)

I have the same concern as with the missed blocks, but less so since these blocks are actually produced and we might want to track bad actors (though we can do that with RPC calling individual blocks of the epoch that we find suspicious)
